File: 파이썬 알고리즘 인터뷰/6장/가장 긴 팰린드롬 부분 문자열.js
// 나의 풀이 (Runtime: 92ms, Memory: 41.3MB)
const expand = (string, left, right) => {
  // left와 right가 범위 내에 있을 때, 양쪽으로 확장하며 팰린드롬 여부를 검사한다.
  while (left >= 0 && right < string.length) {
    if (string[left] !== string[right]) {
      break;
    }

    left--;
    right++;
  }

  // 증감 연산자가 반복문 종료 시 한번 더 계산되므로,  left + 1 해준다. 
  return string.slice(left + 1, right);
}

const longestPalindrome = function(str) {
  // 길이가 1이거나, 문자열 전체가 팰린드롬인 경우 즉시 반환한다.
  if (str.length < 2 || str === str.split("").reverse().join("")) {
    return str;
  }

  let result = "";

  // 짝수, 홀수에 대한 투 포인터 각각을 이동시키며 확인한다.
  for (let idx = 0; idx < str.length; idx++) {
    const evenPalindrome = expand(str, idx, idx + 1); // 2 -> 4 -> 6
    const oddPalindrome = expand(str, idx, idx + 2);  // 3 -> 5 -> 7

    result = evenPalindrome.length > result.length ? evenPalindrome : result;
    result = oddPalindrome.length > result.length ? oddPalindrome : result;
  }

  return result;
};

File: 파이썬 알고리즘 인터뷰/6장/가장 긴 팰린드롬 부분 문자열.py
class Solution:
    # 나의 풀이 (Runtime: 8224ms, Memory: 14.3MB)
    def longestPalindrome(self, s: str) -> str:
        answer = ""

        for i in range(len(s)):
            for j in range(len(s) - 1, i - 1, -1):
                sliced = s[i:j + 1]

                if len(sliced) < len(answer):
                    break

                if sliced == sliced[::-1] and len(answer) < len(sliced):
                    answer = sliced

        return answer

    # 투 포인터/슬라이딩 윈도우를 이용한 풀이 (Runtime: 256ms, Memory: 14MB)
    def answer_with_two_pointers(self, string: str) -> str:
        def expand(left: int, right: int) -> str:
            # 투 포인터가 범위 내에 있을 경우, 팰린드롬 여부를 검사한다.
            while left >= 0 and right < len(string):
                if string[left] == string[right]:
                    left -= 1
                    right += 1
                else:
                    break
            return string[left + 1:right]

        # 문자열의 길이가 1이거나, 문자열 전체가 팰린드롬인 경우 즉시 반환한다.
        if len(string) < 2 or string == string[::-1]:
            return string

        result = ''

        for idx in range(0, len(string) - 1):
            # 짝수 길이, 홀수 길이 팰린드롬 여부를 확인하고 길이가 최대인 문자열을 저장한다.
            result = max(result,
                         expand(idx, idx + 1),
                         expand(idx, idx + 2),
                         key=len)

        return result
